<a href="https://colab.research.google.com/github/AIWalaBro/Complete_NLP/blob/main/NLP_Assignments/1_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1)Reading Text_File

In [71]:
import pandas as pd
pd.set_option('display.max_rows', None)

# Assuming you have a DataFrame named 'df' with a column containing long text
# Set the display.max_colwidth option to a large value to prevent truncation
pd.set_option('display.max_colwidth', None)





In [72]:
import os
import sys
with open('/content/random_generated_text.txt','r') as f:
  text = f.readlines()
  print(text)

['today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"\n', 'and here is his response:-\n', '\n', 'Sure, here is a random news of today which is useful:🥰\n', '\n', 'Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule. The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.\n', 'New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia. The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.\n', 'World leaders agree to phase out coal power: World leaders at the COP26 climate summit in

In [73]:
text

['today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"\n',
 'and here is his response:-\n',
 '\n',
 'Sure, here is a random news of today which is useful:🥰\n',
 '\n',
 'Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule. The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.\n',
 'New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia. The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.\n',
 'World leaders agree to phase out coal power: World leaders at the COP26 climate su

In [74]:
text_lines = text.copy()

In [75]:
# remove '\n' from the text and join in one paragraph.
text_data = "".join(lines.strip () for lines in text_lines)

In [76]:
type(text_data)

str

#2) do the text preprocessing on that
- `1) by python split fuction`
  - a) word tokenizer
  - b) sentense tokenizer
- `2) by NLTK`
  - a) word tokenizer
  - b) sentense tokenizer
  - c) word punctuation
  - d) tweet tokenizer
- `3) by Spacy`
  - a) senetense tokenizer
  - b) Sentencizer


### lets see how our text_data looks
- this is for our own understanding

In [116]:
text_data
# so our text data looks like this

today is 31/08/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"and here is his response:-
Sure, here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule. The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia. The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow have agreed to phase out coal power by the middle of the century. The agreement is a major step forward in the fight against climate change, but it remains to be seen how it will be implemented.First commercial space flight takes off: The first commercial space flight took off from Florida on Wednesday, carrying a crew of three civilians. The flight, which was organized by Virgin Galactic, marked a major milestone in the commercialization of space travel.New study finds benefits of meditation for mental health: A new study published in the journal JAMA Psychiatry has found that meditation can be beneficial for mental health. The study, which looked at data from over 4,000 people, found that meditation was associated with a reduced risk of depression, anxiety, and stress.😭These are just a few of the many news stories that are useful and informative. I hope you find this information helpful.😊😂❤️

In [78]:
#  counting each and every character including spaces
len(text_data)

1871

In [79]:
# text data splitted on the basis of ","
text_data.split(',')

['today is 8/31/2023',
 ' i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"and here is his response:-Sure',
 ' here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California',
 ' San Diego have developed a new way to detect cancer in blood using a single molecule. The new method',
 ' called nanopore sequencing',
 ' is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia. The study',
 ' which looked at data from over 7',
 '000 people',
 ' found that those who were exposed to higher levels of air pollution were more likely to develop dementia.World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow h

In [80]:
len(text_data.split(','))

18

##a) Text into word token

In [81]:
# 1) use python split function and convert into word token
text_data.split()

['today',
 'is',
 '8/31/2023,',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 '"generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful!"and',
 'here',
 'is',
 'his',
 'response:-Sure,',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful:🥰Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood:',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California,',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule.',
 'The',
 'new',
 'method,',
 'called',
 'nanopore',
 'sequencing,',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 '💀cancer.New',
 'study',
 'finds',
 'link',
 'between',
 'air',
 'pollution',
 'and',
 'dem

In [82]:
# it counting each and every word although it consider special character as an word
# it splitted on the basis of white spaces
len(text_data.split())

310

##b) Text_data into Sentenses

In [83]:
text_data.split('.')

['today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"and here is his response:-Sure, here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule',
 ' The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer',
 'New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia',
 ' The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia',
 'World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow have agreed to ph

## find alpha numeric

In [84]:
import re

tokens = re.findall('[\w]+', text_data)
tokens

['today',
 'is',
 '8',
 '31',
 '2023',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 'generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 'and',
 'here',
 'is',
 'his',
 'response',
 'Sure',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 'Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule',
 'The',
 'new',
 'method',
 'called',
 'nanopore',
 'sequencing',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 'cancer',
 'New',
 'study',
 'finds',
 'link',
 'between',
 'air',
 'pollution',
 

# Traditional way to findout special character are present into data.
-  so we will use regex to clean the data from special character.

In [85]:
# Create an empty set to store unique special characters
special_characters = set()

# Iterate through each character in the text_data
for char in text_data:
    # Check if the character is not alphanumeric
    if not char.isalnum():
        # Add it to the set of special characters
        special_characters.add(char)

# Convert the set of special characters to a sorted list
special_characters_list = sorted(list(special_characters))

# Print the list of special characters
print("Special characters in the text_data:")
for char in special_characters_list:
    print(char)


Special characters in the text_data:
 
!
"
,
-
.
/
:
❤
️
💀
😂
😊
😭
🥰


# regex

In [115]:
# Define a regex pattern to match the specified special characters
special_characters_pattern = r'[!"\-,./:❤️💀😂😊😭🥰]'

# Remove the specified special characters from the text_data
cleaned_text_data = re.sub(special_characters_pattern, '', text_data)

# Print the cleaned text data
cleaned_text_data


'today is 8312023 i write a prompt on bard and this data is genrated by googlebard on the generate random news of today which is usefuland here is his responseSure here is a random news of today which is usefulScientists develop new way to detect cancer in blood Scientists at the University of California San Diego have developed a new way to detect cancer in blood using a single molecule The new method called nanopore sequencing is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of cancerNew study finds link between air pollution and dementia A new study published in the journal Nature has found a link between air pollution and dementia The study which looked at data from over 7000 people found that those who were exposed to higher levels of air pollution were more likely to develop dementiaWorld leaders agree to phase out coal power World leaders at the COP26 climate summit in Glasgow have agreed to phase out coal power by the middle of the ce

## observation:-
`a) word token`
-  we splitted the data on the basis of "," and it converted into the word token.
- we calculated the what is the length  of word token so found out it been 310.

``b) sentense token``
- generally we consider a senetense "." on the basis of fullstop at the end of the sentense, so here try to split on the basis of that but the flaw is most of the lines does have "." even though sentense completion. but still we try to split "."

``d) special character removal``
- removal of all the special characters from the text_data except date..8/31/2023
- lets see we have to wrok upon it because this

#2) NLTK

In [93]:
from nltk.tokenize import word_tokenize

In [94]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

# a) Word tokenizer by NLTK

In [95]:
tokens = word_tokenize(text_data)
tokens

['today',
 'is',
 '8/31/2023',
 ',',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 '``',
 'generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 '!',
 '``',
 'and',
 'here',
 'is',
 'his',
 'response',
 ':',
 '-Sure',
 ',',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 ':',
 '🥰Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 ':',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California',
 ',',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule',
 '.',
 'The',
 'new',
 'method',
 ',',
 'called',
 'nanopore',
 'sequencing',
 ',',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 '💀cancer.New',

# Sentence Tokenizer by NLTK

In [96]:
from nltk.tokenize import sent_tokenize

In [97]:
sentense_token = sent_tokenize(text_data)
sentense_token

['today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!',
 '"and here is his response:-Sure, here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule.',
 'The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia.',
 'The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow have agreed to phase

# word punctuation by NLTK
- on the basis of each and evry punctuation it will splitted, even though there is dot it will split and whitespaces also.

In [98]:
from nltk.tokenize import word_tokenize

In [99]:
word_tokenize(text_data)

['today',
 'is',
 '8/31/2023',
 ',',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 '``',
 'generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 '!',
 '``',
 'and',
 'here',
 'is',
 'his',
 'response',
 ':',
 '-Sure',
 ',',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 ':',
 '🥰Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 ':',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California',
 ',',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule',
 '.',
 'The',
 'new',
 'method',
 ',',
 'called',
 'nanopore',
 'sequencing',
 ',',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 '💀cancer.New',

# Tweet Tokenizer

In [100]:
from nltk.tokenize import TweetTokenizer

In [105]:
tokenizer = TweetTokenizer()
tokenizer.tokenize(text_data)

['today',
 'is',
 '8/',
 '31/2023',
 ',',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 '"',
 'generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 '!',
 '"',
 'and',
 'here',
 'is',
 'his',
 'response',
 ':',
 '-',
 'Sure',
 ',',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful',
 ':',
 '🥰',
 'Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 ':',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California',
 ',',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule',
 '.',
 'The',
 'new',
 'method',
 ',',
 'called',
 'nanopore',
 'sequencing',
 ',',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 '

# Spacy

In [106]:
import spacy

# Load the English language model
nlp = spacy.load("en_core_web_sm")


In [108]:
doc = nlp(text_data)
token_list = []
for tokens in doc:
  token_list.append(tokens.text)


token_list

['today',
 'is',
 '8/31/2023',
 ',',
 'i',
 'write',
 'a',
 'prompt',
 'on',
 'bard',
 'and',
 'this',
 'data',
 'is',
 'genrated',
 'by',
 'googlebard',
 'on',
 'the',
 '"',
 'generate',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful!"and',
 'here',
 'is',
 'his',
 'response:-Sure',
 ',',
 'here',
 'is',
 'a',
 'random',
 'news',
 'of',
 'today',
 'which',
 'is',
 'useful:',
 '🥰',
 'Scientists',
 'develop',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 ':',
 'Scientists',
 'at',
 'the',
 'University',
 'of',
 'California',
 ',',
 'San',
 'Diego',
 'have',
 'developed',
 'a',
 'new',
 'way',
 'to',
 'detect',
 'cancer',
 'in',
 'blood',
 'using',
 'a',
 'single',
 'molecule',
 '.',
 'The',
 'new',
 'method',
 ',',
 'called',
 'nanopore',
 'sequencing',
 ',',
 'is',
 'more',
 'sensitive',
 'than',
 'traditional',
 'methods',
 'and',
 'could',
 'lead',
 'to',
 'earlier',
 'diagnosis',
 'and',
 'treatment',
 'of',
 '💀',
 'cancer',
 '.',
 'New',
 'study',
 

# Sentence Tokenizer

In [111]:
doc = nlp(text_data)
sentenses_list = [sentense.text for sentense in doc.sents]
sentenses_list

['today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"and here is his response:-Sure, here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule.',
 'The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.',
 'New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia.',
 'The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.',
 'World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow have agreed to 

#Sentencizer
- with the help of sentensizer function we can convert text into the list of senetenses

In [112]:
from spacy.lang.en import English

In [113]:
nlp = English()
nlp.add_pipe('sentencizer')
# process the text_data

doc = nlp(text_data)

# access the sentence
for sentenses in doc.sents:
  print(sentenses.text)




today is 8/31/2023, i write a prompt on bard and this data is genrated by googlebard on the "generate random news of today which is useful!"and here is his response:-Sure, here is a random news of today which is useful:🥰Scientists develop new way to detect cancer in blood: Scientists at the University of California, San Diego have developed a new way to detect cancer in blood using a single molecule.
The new method, called nanopore sequencing, is more sensitive than traditional methods and could lead to earlier diagnosis and treatment of 💀cancer.
New study finds link between air pollution and dementia: A new study published in the journal Nature has found a link between air pollution and dementia.
The study, which looked at data from over 7,000 people, found that those who were exposed to higher levels of air pollution were more likely to develop dementia.
World leaders agree to phase out coal power: World leaders at the COP26 climate summit in Glasgow have agreed to phase out coal pow